### DISCUSSION TITLE
[Unstable] Speaker/jukebox bugs and inconsistencies

### DISCUSSION THREAD

**Discussion by NotWendy** [FakeFishGames/Barotrauma#14841] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

This is a set of issues I discovered with the new items that play audio, which are used in outpost modules.
- Speakers/jukeboxes which are on at the start of the round begin playing their music slightly out of sync. Annoying when you want to put multiple speakers in one module to even out the volume across a large room.
- Changing the state of the speaker/jukebox with wiring is broken. Speaker has a state checkbox that's visible when you manually interact with the item. Sending a signal to switch the speaker off via wiring does NOT affect the state of the visible checkbox.
In order for the speaker to be on and play music, the visible checkbox needs to be checked on manually AND the invisible state variable needs to be set to 1 via wiring.
Compare this to something like a relay, where sending a toggle_state signal into it toggles the state of its state checkbox. This is the expected behavior.
- Speakers and jukeboxes don't have a power pin, but they do have power consumption specified in the item settings by default. Why? Doesn't seem to affect anything, it's just odd.
- Speakers and jukeboxes play the same tracks, but the order in which tracks are numbered is reversed in the jukebox. It'd be better to keep the numbering consistent.

All testing was done within the submarine editor's test mode.

### Reproduction steps

For the wiring issue:
1. Place a speaker in the outpost module.
2. Press E on it to open the interaction menu and uncheck the "On" checkbox.
3. Wire a switch to the "set_state" pin or a button to the "toggle_state" pin.
4. Load the module in the test mode and interact with the switch/button.
5. The speaker does not play music.
6. Repeat the process with the "On" checkbox being checked.
7. The switch/button now properly affects the speaker's state, but regardless of whether it's wired to be on or off, the "On" checkbox will maintain its original state.

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Happens outside single player or multiplayer game modes (e.g. game launches on startup, something broken in the main menu)

### Version

v1.6.4.0 (unstable)

### Which operating system did you encounter this bug on?

Windows

**Comment by girlyguppy** [comment]
Also they play the exact same tracks as jukebox, just in reverse order

10 second loops why? Give us basically the full song

**Reply by NotWendy** [reply]
I did mention the first part in the post, but the point about playing full tracks is something I fully agree with.
Oh, and having a dynamic system for muffling and eventually fully muting the background music which normally plays at outposts as you get close to an active speaker would be great too.

**Reply by girlyguppy** [reply]
Yeah, music attenuation is an absolute must. We can't have them overlapping

**Comment by Jasontti** [comment]
Tested a fix for this.

Issues addressed:
- Option to make sounds mute the background music when playing. 
- Made speakers and jukeboxes mute background music.
- Fixed "on" tickbox on jukeboxes and speakers not getting refreshed when the item is turned on/off using signals.

Could not repro out of sync issue anymore, so could also be fixed.
Power pin and track ordering still stands, I'll leave this discussion up for those.

**Reply by UnhappyWithMyUsername** [reply]
The out of sync issue is just the fact that the audio clip only starts playing for the character that is in range. So if you just enter the range of speaker A, wait a few seconds and then move slightly to enter the range of speaker B, they'll be out of sync as speaker B just started playing the audio file.
This also causes weird sound issues in MP for example a custom song is ending for player A but player B just got into the range so it's just starting for them.

To fix that the game would need to be able to keep track of time passed and start playing the audio file from the point it would be by that tracked time.
